### PR TITLE
Better room reservation

### PR DESCRIPTION
**Pull request by WhiteHalmos**
- [x] Start reserving from RCL 4.
- [x] Reserve only one room at RCL 4 (3 is way too much).
- [x] Only reserve next to own room or own reserved room path.

### CODE REVIEW

**Comment by WhiteHalmos** [comment]
Ready for review. :)

File: src/prototype_room_external.js [hunk 1]
```diff
@@ -296,24 +294,31 @@ Room.prototype.handleReservedRoom = function() {
...
       if (!this.memory.reservation) {
         this.log('No reservation');
-        return;
+        return false;
       }
 
-      if (Game.rooms[this.memory.reservation.base].misplacedSpawn) {
-        energyThreshold = 1600;
+      let reservation = this.memory.reservation;
+      let energyNeeded = 1300;
```
Comment (TooAngel) [inline src/prototype_room_external.js]: Maybe this can be extracted to a `config` value and `https://github.com/TooAngel/screeps/blob/master/src/prototype_room_utils.js#L3` to check the misplacedSpawn thingy,

File: src/prototype_room_external.js [hunk 2]
```diff
@@ -355,14 +362,18 @@ Room.prototype.handleUnreservedRoom = function() {
           }
         };
         // TODO move the creep check from the reserver to here and spawn only sourcer (or one part reserver) when controller.level < 4
-        let energyThreshold = 1300;
-        if (Game.rooms[this.memory.reservation.base].misplacedSpawn) {
-          energyThreshold = 1600;
+        let energyNeeded = 1300;
```
Comment (TooAngel) [inline src/prototype_room_external.js]: Same as before, extract to `config` use misplacedSpawn method

File: src/prototype_room_external.js [hunk 3]
```diff
@@ -390,26 +401,34 @@ Room.prototype.handleUnreservedRoom = function() {
...
-      if (reservedRooms.length < room.controller.level - 1) {
+      /* RCL: target reserved rooms
+       * 4: 1
+       * 5: 3
+       * 6: 5
+       * 7: 7
+       * 8: 9
+       */
+      let numRoomsToReserve = (room.controller.level - 3) * 2 - 1;
+      if (reservedRooms.length < numRoomsToReserve) {
```
Comment (TooAngel) [inline src/prototype_room_external.js]: The formula was recently introduced to sure if it is fine with the number of external rooms.

Currently I'm more thinking that the number of external rooms should depend on the number of spawns.
1 Spawn (RCL 1-6): 3 rooms
2 Spawns (RCL 7): 6 rooms
3 Spawns (RCL 8): 9 rooms

(Replace the number of rooms, with good numbers :-))

Comment (TooAngel) [inline src/prototype_room_external.js]: I'm fine with your change, just want to give some idea for improvement

File: src/prototype_room_external.js [hunk 4]
```diff
@@ -422,20 +441,20 @@ Room.prototype.handleUnreservedRoom = function() {
           level: 2
         };
         // TODO move the creep check from the reserver to here and spawn only sourcer (or one part reserver) when controller.level < 4
-        let energyThreshold = 1300;
-        if (Game.rooms[this.memory.reservation.base].misplacedSpawn) {
-          energyThreshold = 1600;
+        let energyNeeded = 1300;
```
Comment (TooAngel) [inline src/prototype_room_external.js]: same again `config` ...

**Comment by WhiteHalmos** [comment]
Please take a look. Not sure why circleci is failing though. :(

**Comment by TooAngel** [comment]
If you click on `Details` you should get redirected to CircleCI showing you the error:

```
Running "jshint:all" (jshint) task

   src/prototype_room_external.js
    378 |        let reservedRooms = _.filter(Memory.rooms, (roomMemory) => {
                                                            ^ Don't make functions within a loop.
```

File: src/prototype_room_external.js [hunk 5]
```diff
@@ -154,6 +146,12 @@ Room.prototype.externalHandleHighwayRoom = function() {
       Memory.powerBanks = {};
     }
     if (target !== null) {
+      // Fix for W8S9
```
Comment (TooAngel) [inline src/prototype_room_external.js]: I think this part could even be removed. IIRC this was about handling newbie walls. Which never worked properly.

Also fine to leave it in.

File: src/prototype_room_external.js [hunk 6]
```diff
@@ -268,174 +321,98 @@ Room.prototype.handleReservedRoom = function() {
...
-            routePos: 0,
-            pathPos: 0
-          }
-        };
-        // TODO move the creep check from the reserver to here and spawn only sourcer (or one part reserver) when controller.level < 4
-        let energyThreshold = 1300;
-        if (Game.rooms[this.memory.reservation.base].misplacedSpawn) {
-          energyThreshold = 1600;
+  if (this.memory.reservation === undefined) {
+    checkRoomsLabel: for (let roomName of Memory.myRooms) {
```
Comment (TooAngel) [inline src/prototype_room_external.js]: Didn't know the label functionality, nice.

**Comment by WhiteHalmos** [comment]
I like that idea of mining without reservation. If you like the code as is here, I'd like to do that as a separate feature/pull-request. This one is getting a bit long. :)

**Comment by TooAngel** [comment]
> I like that idea of mining without reservation. If you like the code as is here, I'd like to do that as a separate feature/pull-request. This one is getting a bit long. :)

Yes, was just a hint if you like to improve external harvesting further. Not meant for this PR

File: src/prototype_room_external.js [hunk 7]
```diff
@@ -260,182 +238,169 @@ Room.prototype.checkBlockedPath = function() {
...
+        let numRooms = {
+          0: 0,
+          1: 0,
+          2: 0,
+          3: 0,
+          4: 1,
+          5: 2,
+          6: 3,
+          7: 6,
+          8: 9,
```
Comment (TooAngel) [inline src/prototype_room_external.js]: Didn't recognize this on the last PR. I would merge this now.

But: :-) Maybe better in the config or I like usually if it is calculable.

Comment (WhiteHalmos) [inline src/prototype_room_external.js]: Sounds good, will pull out into config in another PR. :)
